### ISSUE TITLE
Fix S2589 FP: `NotNullWhen` leads to unexpected behavior

### ISSUE THREAD

**Issue by Tim-Pohlmann** [SonarSource/sonar-dotnet#9051]
S2589 ignores literal checks. Thus, it should not be raised here:
```cs
using System;
using System.Diagnostics.CodeAnalysis;
class Test
{
    string Method(bool b) => TrySomething(out var s) switch
    {
        true => s,
        false => ""  // Noncompliant FP
    };

    bool TrySomething([NotNullWhen(true)] out string outValue) => throw new NotImplementedException();
}
```
If you remove `NotNullWhen`, the issue goes away.

**Comment by martin-strecker-sonarsource** [comment]
Tracked by https://sonarsource.atlassian.net/browse/NET-1213 now
